enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,603 @@
+---
+title: set-replace-append-delete-deny-headers
+authors:
+  - "@miheer"
+reviewers:
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+approvers:
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+api-approvers:
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+creation-date: 2022-12-02
+last-updated: 2022-12-14
+tracking-link:
+  - "https://issues.redhat.com/browse/NE-982"
+status: TODO
+see-also:
+- ""
+replaces:
+superseded-by:
+---
+
+# Set(Replace), Append, Delete, Deny HTTP Response headers
+
+## Summary
+
+This enhancement extends the IngressController API and Route API to allow the user to set(replace) or append 
+HTTP Response headers in the haproxy.config file. Using the new API, users may specify a list of 
+HTTP Response headers either via IngressController API or Route API that should be set(replace) or appended or deleted.
+Creates an immutable(hardcoded) list of HTTP Response headers which can't be set(replace) or appended or deleted.
+
+## Motivation
+
+Certain HTTP response headers may need to be set(replace), appended, deleted for compliance reasons or other reasons. In OpenShift 3, cluster 
+administrators could use a custom HAProxy configuration template to set(replace)/append headers, but using a custom
+template is not permitted in OpenShift 4.
+Certain HTTP Response headers like HSTS(Strict-Transport-Security) need to denied as we already provide a way to set them
+via route annotation.
+
+### User Stories
+
+1. As a cluster administrator, I need to configure an IngressController to set(replace)/append certain HTTP Response headers which are:
+   - X-XSS-Protection "1; mode=block"
+   - X-Content-Type-Options "nosniff"
+   - X-Frame-Options "SAMEORIGIN"
+   in order to satisfy compliance requirements at global level to send them to all clients.
+2. As a user/devops/developer, I need to configure Route to set/append HTTP Response headers which need to  which 
+   are mentioned above to send them to the client which has called the route.
+3. To satisfy this use-case, the cluster administrator can set the IngressController's Spec 
+   or the developer can set Router's `spec.httpDeleteHeaders` field to set(replace) or append the required headers.
+4. As a cluster administrator, I need to configure an IngressController to delete certain HTTP Response headers which are:
+   - X-XSS-Protection
+   - X-Content-Type-Options
+   - X-Frame-Options
+     in order to satisfy compliance requirements at global level to not to send them to all clients.
+5. As a user/devops/developer, I need to configure Route to delete HTTP Response headers which need to  which
+      are mentioned above for not sending them to the client which has called the route.
+6. To satisfy this use-case, the cluster administrator can set the IngressController's Spec
+   or the developer can set the Router's Spec with `spec.httpDeleteHeaders` field to delete the required headers by 
+   mentioning the header names mentioned above.
```

Comment:
Is there precedent for writing, "As a developer, I need to update ___ API" as a user story?  User stories are usually focused on end-user roles, such as cluster admin or project admin.  

I do like simplifying the user stories where possible and simplifying the list so that it enumerates one user story per list item.  It is helpful to provide the solution to the user story though.  When writing an EP, I put the user story as a heading or italicized statement followed by one or more paragraphs explaining how the EP addresses the user story.  

The user story should indicate who the end-user is (e.g. "cluster admin"), what the end user is trying to achieve (e.g. "delete the `x-server` response header in responses to client requests") and why ("so that I can avoid exposing information about internal systems to potentially malicious external actors").  Then you can explain how the EP solves that user story ("The cluster admin can configure the IngressController as shown in the following example yaml").

Reply:
> Is there precedent for writing, "As a developer, I need to update ___ API" as a user story? User stories are usually focused on end-user roles, such as cluster admin or project admin.

I created such user stories myself, for technical debt for instance. I suppose it's in the interest of NetEdge engineer to define and implement the deny list. That's why I put Developer as the actor for the first user story.

> The user story should indicate who the end-user is (e.g. "cluster admin"), what the end user is trying to achieve (e.g. "delete the x-server response header in responses to client requests") and why ("so that I can avoid exposing information about internal systems to potentially malicious external actors"). Then you can explain how the EP solves that user story ("The cluster admin can configure the IngressController as shown in the following example yaml").

Isn't the Workflow chapter for this purpose?

Reply:
> > Is there precedent for writing, "As a developer, I need to update ___ API" as a user story? User stories are usually focused on end-user roles, such as cluster admin or project admin.
> 
> I created such user stories myself, for technical debt for instance. I suppose it's in the interest of NetEdge engineer to define and implement the deny list. That's why I put Developer as the actor for the first user story.

I can see creating user stories in Jira for tech debt, and we sometimes note potential future follow-up work in the goals/non-goals, proposal, or implementation details/notes/constraints sections.  However, I still believe that user stories in the enhancement should be focused on end users.  

> > The user story should indicate who the end-user is (e.g. "cluster admin"), what the end user is trying to achieve (e.g. "delete the x-server response header in responses to client requests") and why ("so that I can avoid exposing information about internal systems to potentially malicious external actors"). Then you can explain how the EP solves that user story ("The cluster admin can configure the IngressController as shown in the following example yaml").
> 
> Isn't the Workflow chapter for this purpose?

I would use user stories for the high-level description of the problem ("As a ___, I want to ___, so that I can ___") and solution ("To accomplish ___, ___ can use the new ___ API/field/whatever as in the following example") and then document the detailed step-by-step procedure as a workflow.

Reply:
done